File: examples/python/hls_packager.py
"""
FFmpeg HLS Packaging Example in Python
=======================================
Generates a multi-bitrate HLS package from a single source file.

Usage:
    python hls_packager.py input.mp4 ./output_hls

Requirements:
    - ffmpeg with libx264 and fdk_aac/aac support
    - Python 3.10+
"""

from __future__ import annotations

import argparse
import os
import subprocess
import sys
from dataclasses import dataclass


@dataclass
class RenditionConfig:
    """Defines a single HLS rendition (resolution + bit rate)."""

    name: str
    width: int
    height: int
    video_bitrate_kbps: int
    audio_bitrate_kbps: int
    max_rate_multiplier: float = 1.07


# Default multi-bitrate ladder
DEFAULT_RENDITIONS: list[RenditionConfig] = [
    RenditionConfig("1080p", 1920, 1080, 4500, 192),
    RenditionConfig("720p",  1280,  720, 2500, 128),
    RenditionConfig("480p",   854,  480, 1000,  96),
    RenditionConfig("360p",   640,  360,  500,  64),
]


def build_hls_command(
    input_path: str,
    output_dir: str,
    renditions: list[RenditionConfig],
    segment_duration: int = 6,
    use_fmp4: bool = True,
    keyframe_interval: int = 60,
) -> list[str]:
    """
    Build the FFmpeg command for multi-bitrate HLS packaging.

    Args:
        input_path: Source video file.
        output_dir: Output directory for HLS segments and playlists.
        renditions: List of renditions to produce.
        segment_duration: HLS segment duration in seconds.
        use_fmp4: Use fragmented MP4 segments (recommended) vs. MPEG-TS.
        keyframe_interval: Fixed keyframe interval (should be multiple of segment_duration * fps).

    Returns:
        FFmpeg command as a list of strings.
    """
    n = len(renditions)

    # Build the video split filter
    split_inputs = "".join(f"[v{i}]" for i in range(n))
    filter_parts = [f"[0:v]split={n}{split_inputs}"]

    for i, r in enumerate(renditions):
        filter_parts.append(f"[v{i}]scale={r.width}:{r.height}[vout{i}]")

    filter_complex = "; ".join(filter_parts)

    cmd = ["ffmpeg", "-y", "-i", input_path, "-filter_complex", filter_complex]

    # Video streams
    for i, r in enumerate(renditions):
        bv = f"{r.video_bitrate_kbps}k"
        maxrate = f"{int(r.video_bitrate_kbps * r.max_rate_multiplier)}k"
        bufsize = f"{r.video_bitrate_kbps * 2}k"
        cmd += [
            "-map", f"[vout{i}]",
            f"-c:v:{i}", "libx264",
            f"-b:v:{i}", bv,
            f"-maxrate:v:{i}", maxrate,
            f"-bufsize:v:{i}", bufsize,
            f"-preset:v:{i}", "fast",
        ]

    # Audio streams (one per rendition)
    for i, r in enumerate(renditions):
        cmd += [
            "-map", "0:a",
            f"-c:a:{i}", "aac",
            f"-b:a:{i}", f"{r.audio_bitrate_kbps}k",
            f"-ar:{i}", "48000",
        ]

    # Global x264 options for constant keyframe interval
    cmd += [
        "-x264opts", f"keyint={keyframe_interval}:min-keyint={keyframe_interval}:no-scenecut",
        "-g", str(keyframe_interval),
        "-keyint_min", str(keyframe_interval),
        "-sc_threshold", "0",
    ]

    # HLS output options
    segment_ext = "mp4" if use_fmp4 else "ts"
    segment_type_flag = ["-hls_segment_type", "fmp4"] if use_fmp4 else []

    master_pl = os.path.join(output_dir, "master.m3u8")
    var_stream_map = " ".join(f"v:{i},a:{i}" for i in range(n))
    segment_filename = os.path.join(output_dir, "stream_%v", f"seg_%03d.{segment_ext}")
    variant_playlist = os.path.join(output_dir, "stream_%v", "playlist.m3u8")

    cmd += [
        "-f", "hls",
        "-hls_time", str(segment_duration),
        "-hls_playlist_type", "vod",
        "-hls_flags", "independent_segments",
        *segment_type_flag,
        "-master_pl_name", master_pl,
        "-var_stream_map", var_stream_map,
        "-hls_segment_filename", segment_filename,
        variant_playlist,
    ]

    return cmd


def package_hls(
    input_path: str,
    output_dir: str,
    renditions: list[RenditionConfig] | None = None,
    segment_duration: int = 6,
) -> None:
    """
    Package a video file as multi-bitrate HLS.

    Args:
        input_path: Source video.
        output_dir: Directory to write HLS output.
        renditions: Custom rendition configurations (default ladder used if None).
        segment_duration: Duration of each HLS segment in seconds.
    """
    if renditions is None:
        renditions = DEFAULT_RENDITIONS

    # Create output subdirectories for each rendition
    for i in range(len(renditions)):
        os.makedirs(os.path.join(output_dir, f"stream_{i}"), exist_ok=True)

    cmd = build_hls_command(
        input_path,
        output_dir,
        renditions,
        segment_duration=segment_duration,
    )

    print(f"Packaging HLS with {len(renditions)} renditions...")
    print(f"Command: {' '.join(cmd)}\n")

    subprocess.run(cmd, check=True)
    print(f"\nHLS output written to: {output_dir}")
    print(f"Master playlist: {os.path.join(output_dir, 'master.m3u8')}")


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Package a video file as multi-bitrate HLS",
    )
    parser.add_argument("input", help="Input video file path")
    parser.add_argument("output", help="Output directory for HLS package")
    parser.add_argument(
        "--segment-duration", type=int, default=6,
        help="HLS segment duration in seconds (default: 6)",
    )
    args = parser.parse_args()

    if not os.path.isfile(args.input):
        print(f"Error: input file not found: {args.input}", file=sys.stderr)
        sys.exit(1)

    package_hls(args.input, args.output, segment_duration=args.segment_duration)


if __name__ == "__main__":
    main()

File: examples/nodejs/hls-packager.js
/**
 * FFmpeg HLS Packager for Node.js
 * ================================
 * Generates a multi-bitrate HLS package from a single source video.
 *
 * Usage:
 *   node hls-packager.js input.mp4 ./output_hls
 *   node hls-packager.js input.mp4 ./output_hls --segment-duration 4
 *
 * Requirements:
 *   - ffmpeg with libx264 support
 *   - Node.js 16+
 */

'use strict';

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// ---------------------------------------------------------------------------
// Default ABR ladder configuration
// ---------------------------------------------------------------------------

/**
 * @typedef {object} Rendition
 * @property {string} name - Rendition name (e.g., "1080p")
 * @property {number} width - Video width
 * @property {number} height - Video height
 * @property {number} videoBitrateKbps - Video bit rate in kbps
 * @property {number} audioBitrateKbps - Audio bit rate in kbps
 */

/** @type {Rendition[]} */
const DEFAULT_RENDITIONS = [
  { name: '1080p', width: 1920, height: 1080, videoBitrateKbps: 4500, audioBitrateKbps: 192 },
  { name: '720p',  width: 1280, height: 720,  videoBitrateKbps: 2500, audioBitrateKbps: 128 },
  { name: '480p',  width: 854,  height: 480,  videoBitrateKbps: 1000, audioBitrateKbps: 96  },
  { name: '360p',  width: 640,  height: 360,  videoBitrateKbps: 500,  audioBitrateKbps: 64  },
];

// ---------------------------------------------------------------------------
// Command builder
// ---------------------------------------------------------------------------

/**
 * Build the FFmpeg argument list for multi-bitrate HLS packaging.
 *
 * @param {object} params
 * @param {string} params.inputPath - Source video file.
 * @param {string} params.outputDir - Directory for HLS output.
 * @param {Rendition[]} params.renditions - Array of rendition configs.
 * @param {number} [params.segmentDuration=6] - HLS segment duration in seconds.
 * @param {number} [params.keyframeInterval=60] - Keyframe interval (frames).
 * @param {boolean} [params.useFmp4=true] - Use fMP4 segments (vs MPEG-TS).
 * @returns {string[]} FFmpeg arguments.
 */
function buildHlsArgs({
  inputPath,
  outputDir,
  renditions,
  segmentDuration = 6,
  keyframeInterval = 60,
  useFmp4 = true,
}) {
  const n = renditions.length;

  // Video split + scale filter
  const splitPads = renditions.map((_, i) => `[v${i}]`).join('');
  const filterParts = [`[0:v]split=${n}${splitPads}`];
  renditions.forEach((r, i) => {
    filterParts.push(`[v${i}]scale=${r.width}:${r.height}[vout${i}]`);
  });
  const filterComplex = filterParts.join('; ');

  const args = ['-y', '-i', inputPath, '-filter_complex', filterComplex];

  // Per-rendition video encoding options
  renditions.forEach((r, i) => {
    const bv = `${r.videoBitrateKbps}k`;
    const maxrate = `${Math.round(r.videoBitrateKbps * 1.07)}k`;
    const bufsize = `${r.videoBitrateKbps * 2}k`;
    args.push(
      '-map', `[vout${i}]`,
      `-c:v:${i}`, 'libx264',
      `-b:v:${i}`, bv,
      `-maxrate:v:${i}`, maxrate,
      `-bufsize:v:${i}`, bufsize,
      `-preset:v:${i}`, 'fast',
    );
  });

  // Per-rendition audio encoding options
  renditions.forEach((r, i) => {
    args.push(
      '-map', '0:a',
      `-c:a:${i}`, 'aac',
      `-b:a:${i}`, `${r.audioBitrateKbps}k`,
      `-ar:${i}`, '48000',
    );
  });

  // Fixed keyframe interval (critical for HLS segment alignment)
  args.push(
    '-x264opts', `keyint=${keyframeInterval}:min-keyint=${keyframeInterval}:no-scenecut`,
    '-g', String(keyframeInterval),
    '-keyint_min', String(keyframeInterval),
    '-sc_threshold', '0',
  );

  // HLS output options
  const segExt = useFmp4 ? 'mp4' : 'ts';
  const masterPl = path.join(outputDir, 'master.m3u8');
  const varStreamMap = renditions.map((_, i) => `v:${i},a:${i}`).join(' ');
  const segmentFilename = path.join(outputDir, 'stream_%v', `seg_%03d.${segExt}`);
  const variantPlaylist = path.join(outputDir, 'stream_%v', 'playlist.m3u8');

  args.push(
    '-f', 'hls',
    '-hls_time', String(segmentDuration),
    '-hls_playlist_type', 'vod',
    '-hls_flags', 'independent_segments',
  );

  if (useFmp4) {
    args.push('-hls_segment_type', 'fmp4');
  }

  args.push(
    '-master_pl_name', masterPl,
    '-var_stream_map', varStreamMap,
    '-hls_segment_filename', segmentFilename,
    variantPlaylist,
  );

  return args;
}

// ---------------------------------------------------------------------------
// HLS packager
// ---------------------------------------------------------------------------

/**
 * Package a video as multi-bitrate HLS.
 *
 * @param {object} params
 * @param {string} params.inputPath - Source video.
 * @param {string} params.outputDir - Output directory.
 * @param {Rendition[]} [params.renditions] - Custom renditions (default ladder if omitted).
 * @param {number} [params.segmentDuration=6] - HLS segment duration.
 * @param {Function} [params.onLog] - Called with each FFmpeg stderr line.
 * @returns {Promise<void>}
 */
function packageHls({
  inputPath,
  outputDir,
  renditions = DEFAULT_RENDITIONS,
  segmentDuration = 6,
  onLog = null,
}) {
  // Create rendition subdirectories
  renditions.forEach((_, i) => {
    fs.mkdirSync(path.join(outputDir, `stream_${i}`), { recursive: true });
  });

  const args = buildHlsArgs({ inputPath, outputDir, renditions, segmentDuration });

  console.log(`Packaging HLS with ${renditions.length} renditions...`);
  console.log(`Renditions: ${renditions.map(r => r.name).join(', ')}`);

  return new Promise((resolve, reject) => {
    const proc = spawn('ffmpeg', args, { stdio: ['pipe', 'pipe', 'pipe'] });

    proc.stderr.on('data', (data) => {
      const line = data.toString();
      if (onLog) {
        onLog(line);
      } else {
        process.stderr.write(line);
      }
    });

    proc.on('close', (code) => {
      if (code === 0) {
        console.log(`\nHLS output: ${outputDir}`);
        console.log(`Master playlist: ${path.join(outputDir, 'master.m3u8')}`);
        resolve();
      } else {
        reject(new Error(`FFmpeg exited with code ${code}`));
      }
    });

    proc.on('error', reject);
  });
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

if (require.main === module) {
  const args = process.argv.slice(2);

  const inputPath = args[0];
  const outputDir = args[1];

  if (!inputPath || !outputDir) {
    console.error('Usage: node hls-packager.js <input> <output_dir> [--segment-duration N]');
    process.exit(1);
  }

  if (!fs.existsSync(inputPath)) {
    console.error(`Input file not found: ${inputPath}`);
    process.exit(1);
  }

  const segIdx = args.indexOf('--segment-duration');
  const segmentDuration = segIdx >= 0 ? parseInt(args[segIdx + 1], 10) : 6;

  packageHls({ inputPath, outputDir, segmentDuration })
    .then(() => console.log('Done.'))
    .catch((err) => {
      console.error('Error:', err.message);
      process.exit(1);
    });
}

module.exports = { packageHls, buildHlsArgs, DEFAULT_RENDITIONS };
